src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -493,411 +493,996 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
+            // the end and forwards, which may overlap on an earlier compare.
 
-                    while (length > 0)
+            // We include the Supported check again here even though path will not be taken, so the asm isn't generated if not supported.
+            if (Sse2.IsSupported)
+            {
+                int matches;
+                if (Avx2.IsSupported)
+                {
+                    Vector256<ushort> search;
+                    // Guard as we may only have a valid size for Vector128; when we will move to the Sse2
+                    // We have already subtracted Vector128<ushort>.Count from lengthToExamine so compare against that
+                    // to see if we have double the size for Vector256<ushort>.Count
+                    if (lengthToExamine >= (nuint)Vector128<ushort>.Count)
                     {
-                        // Using Unsafe.Read instead of ReadUnaligned since the search space is pinned and pCh is always vector aligned
-                        Debug.Assert(((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) == 0);
-                        Vector<ushort> vData = Unsafe.Read<Vector<ushort>>(pCh);
-                        var vMatches = Vector.BitwiseOr(
-                                            Vector.BitwiseOr(
-                                                Vector.BitwiseOr(Vector.Equals(vData, values0), Vector.Equals(vData, values1)),
-                                                Vector.Equals(vData, values2)),
-                                            Vector.Equals(vData, values3));
+                        Vector256<ushort> values0 = Vector256.Create(value0);
+                        Vector256<ushort> values1 = Vector256.Create(value1);
+                        Vector256<ushort> values2 = Vector256.Create(value2);
+                        Vector256<ushort> values3 = Vector256.Create(value3);
+
+                        // Subtract Vector128<ushort>.Count so we have now subtracted Vector256<ushort>.Count
+                        lengthToExamine -= (nuint)Vector128<ushort>.Count;
+                        // First time this checks again against 0, however we will move into final compare if it fails.
+                        while (lengthToExamine > offset)
+                        {
+                            search = LoadVector256(ref searchStart, offset);
+                            // We preform the Or at non-Vector level as we are using the maximum number of non-preserved registers,
+                            // and more causes them first to be pushed to stack and then popped on exit to preseve their values.
+                            matches = Avx2.MoveMask(Avx2.CompareEqual(values0, search).AsByte());
+                            // Bitwise Or to combine the flagged matches for the second, third and fourth values to our match flags
+                            matches |= Avx2.MoveMask(Avx2.CompareEqual(values1, search).AsByte());
```

Comment:
Using the `Sse2`/`Avx2` Ors is cheaper because they reduce the `MoveMask`s which are more expensive; so I do that in the methods with smaller numbers of parameters; however for the larger parameter methods it pushes the used vector register counts higher than the maximum number of non-preserved registers, which causes the method to start pushing and poping vectors from the stack, which then becomes expensive in itself.

Reply:
[x64 calling convention](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019#callercallee-saved-registers)

> ### Caller/callee saved registers
>
> The x64 ABI considers the registers RAX, RCX, RDX, R8, R9, R10, R11, and **XMM0-XMM5 volatile**. When present, the upper portions of YMM0-YMM15 and ZMM0-ZMM15 are also volatile. On AVX512VL, the ZMM, YMM, and XMM registers 16-31 are also volatile. Consider volatile registers destroyed on function calls unless otherwise safety-provable by analysis such as whole program optimization.
>
> The x64 ABI considers registers RBX, RBP, RDI, RSI, RSP, R12, R13, R14, R15, and **XMM6-XMM15 nonvolatile**. They must be saved and restored by a function that uses them.

So when we start hitting `XMM6`+ the vectors values start being saved and restored to stack in the prolog/epilog; which then even hits when there is only a single `char` to look at.

However rereading that `YMM0-YMM15 and ZMM0-ZMM15 are also volatile`; should be able to use the Avx2.Or for all of them, is just Sse2 that need to be careful?

Reply:
Yeah, I was check for both SSE2 and Avx2 cases. Did you get chance to do a diff of entire method with `|` vs. `Avx2/Sse2.Or()`?

Reply:
Alas just using more `ymm` without increasing `xmm` usage does still cause `xmm` to get pushed to stack, perhaps because `ymm7` shares half with `xmm7` so still needs to save it?

![image](https://user-images.githubusercontent.com/1142958/90293770-c0ca0f00-de7c-11ea-831f-0978658d2970.png)

Reply:
So are you saying it doesn't make any difference with `|` vs. `Sse2/Avx2.Or()`?

Reply:
So it uses `Sse2/Avx2.Or()` up to 3 values

`IndexOfAny(.., char value0, char value1, ..)`
`IndexOfAny(.., char value0, char value1, char value2, ..)`

At 4 it switches to `|` because doing the ors in Simd requires an extra simd register, and doesn't matter if its just `ymm` rather than `xmm`; so in `IndexOfAny(.., char value0, char value1, char value2, char value3, ..)`

Changing
```csharp
search = LoadVector256(ref searchStart, offset);
// We preform the Or at non-Vector level as we are using the maximum number of non-preserved registers,
// and more causes them first to be pushed to stack and then popped on exit to preseve their values.
matches = Avx2.MoveMask(Avx2.CompareEqual(values0, search).AsByte());
// Bitwise Or to combine the flagged matches for the second, third and fourth values to our match flags
matches |= Avx2.MoveMask(Avx2.CompareEqual(values1, search).AsByte());
matches |= Avx2.MoveMask(Avx2.CompareEqual(values2, search).AsByte());
matches |= Avx2.MoveMask(Avx2.CompareEqual(values3, search).AsByte());
``` 
To
```csharp
search = LoadVector256(ref searchStart, offset);
matches = Avx2.MoveMask(
            Avx2.Or(
                Avx2.CompareEqual(values0, search),
                Avx2.Or(
                    Avx2.CompareEqual(values1, search),
                    Avx2.Or(
                        Avx2.CompareEqual(values2, search),
                        Avx2.CompareEqual(values3, search))
                    )
                ).AsByte());
```
Causes an additional 16 byte push in prologue and 16 byte pop in epilog; which then becomes an additional cost paid even when the vectorized path is not taken (because the lengths are short), when it becomes a significant cost relative to the rest of the `<` 16 byte length path.

Specifically (before on left, after on right):

![image](https://user-images.githubusercontent.com/1142958/90300248-77d08580-de91-11ea-8f22-0a1b69a62226.png)

![image](https://user-images.githubusercontent.com/1142958/90300264-85860b00-de91-11ea-99f9-65a07fe170ee.png)

![image](https://user-images.githubusercontent.com/1142958/90300272-8e76dc80-de91-11ea-8087-7e95fce03778.png)

![image](https://user-images.githubusercontent.com/1142958/90300277-96cf1780-de91-11ea-9fb1-03b0a18eb2f5.png)